### PR TITLE
Feat | Score rules

### LINKED ISSUE #34: configure tasks's deadlines, bonuses & penalties via score rules
Usually every task has a deadline to submit and after that some penalty should be applied.

What is exactly required:
- ability to setup deadline & penalty size by each task individually via config file
- apply penalties to task's score automatically in case if the submission was after a deadline
- export tasks with deadlines & penalties into csv file correctly

_NB: Teacher should not take care about penalty points. For example, student received max points for the task (like 8 points for lab3), but the submission was after a deadline (penalty is 1 point) so the total score would be 8 - 1 = 7 points._

### PR DESCRIPTION
**Pull request by TheGeniusOfEternity**
By this pull request new extra points system is introduced - _Score Rules_. There is new mechanics, score rules, which allows admin to assign bonuses & penalties for specific tasks on specific conditions with specific amount of points.

Resolves #34 , where detailed info might be found

### CODE REVIEW

**Comment by TheGeniusOfEternity** [comment]
Updated _User Profile_ page, added score rules section:

Снимок экрана — 2026-04-21 в 16 51 14

---
How it works:
1) Rule - name of the rule
2) Tasks - names of tasks, related to this rule
3) Condition - how to get/avoid effect of the score rule. There some types:
     - ```after```: ANY TASK of related must be submitted __after__ the specified date to make effect be applied (i.e. if you submit lab1 after the 29.3.2026, you get penalty effect by -2 points)
     - ```before```: ANY TASK of related must be submitted __before__ the specified date to make effect be applied (i.e. if you submit lab4 before the date, you get bonus effect by +1 point)
     - ```after + before```: ANY TASK of related must be submitted __before__ specified date1 __and after__ specified date2 to make effect be applied (i.e. if you submit lab1 after the 15.3.2026 and before the 29.3.2026, you get penalty effect by -1 point)
     - ```less than <value> + before```: works like ```before```, but with extra condition of__amount__ of submitted tasks (i.e. if you submit only lab3.acc32 and lab3.f32  (less than 4) by the end of the semester, you get penalty effect by -2 points)
4) Status:
     - ```applied``` - displayed if rules's condition can't be changed and equals true. Green for ```before``` rules and red for ```after``` rules
     - ```not applied``` - displayed if rules's condition can't be changed and equals false. Gray for all conditions types
     - ```active``` - displayed if rule's condition can be changed.
5) Bonus/Penalty (Effect) - amount of points that would be added if the rule is applied. Color duplicates status

Also above the table there is ```total effect``` - sum of applied points from score rules

**Comment by TheGeniusOfEternity** [comment]
Updated _Users_ page - added _Bonus/Penalty_ column
Снимок экрана — 2026-04-21 в 17 52 49

How it works:
For every student total effect from all applied score rules is collected and added to the table. The column is also included into CSV file on export

File: src/config/config.go [hunk 1]
```diff
@@ -37,6 +43,16 @@ type Task struct {
 	WaitingPeriod *time.Duration `yaml:"waiting_period,omitempty"`
 }
 
+// IsTeacher checks if the given user ID is in the teacher list
```
Comment (ryukzak) [inline src/config/config.go]: useless comment

File: docker-compose.yaml [hunk 2]
```diff
@@ -6,6 +6,9 @@ services:
     volumes:
       - slap-data:/app/tmp
       - ./conf:/app/conf:ro
+    environment:
+      - SLAP_JWT_SECRET=test
+      - SLAP_CONF=conf/config-test.yaml
```
Comment (ryukzak) [inline docker-compose.yaml]: broken config because config is not provided for test. Add it to regular example

File: src/handlers/user.go [hunk 3]
```diff
@@ -73,9 +80,162 @@ func UserInfoHandler(w http.ResponseWriter, r *http.Request) {
 		}
 	}
 
+	var rulesWithStatus []ScoreRuleWithStatus
+	ruleApplies := make(map[string]bool)
+	totalEffect := 0
+
+	if dbUser.IsStudent {
```
Comment (ryukzak) [inline src/handlers/user.go:86]: Looks like we need a simple function for running rules in front of a student or task list or task evaluation list. Right now, it is very complicated. I don’t know about colors.

File: src/storage/db.go [hunk 4]
```diff
@@ -55,3 +55,22 @@ func NewDB(dbPath string, bucketName string) (*DB, error) {
...
+func (d *DB) GetCompletedTasksCountInGroup(userID string, taskIDs []TaskID) (int, error) {
+	completed := 0
+	for _, taskID := range taskIDs {
+		records, err := d.ListTaskRecords(userID, taskID)
+		if err != nil {
+			continue
+		}
+
+		for _, record := range records {
+			if record.Status == ReviewedTaskRecord {
```
Comment (ryukzak) [inline src/storage/db.go]: If I don't mistake -- we need to calculate only the last record. Finished means finished, without upgrades after.

**Review by ryukzak — CHANGES_REQUESTED** [review]
And also we need a tests.

**Review by ryukzak — COMMENTED** [review]
## PR Review: Score Rules

### Critical Issues

**1. Duplicate logic: `CalculateScoreEffects` and `CalculateTotalEffect` are identical**

`config.go` has two methods that do the exact same thing. Only `CalculateTotalEffect` is used. `CalculateScoreEffects` should be removed.

**2. Duplicate rule evaluation logic between `config.go` and `scorerule.go`**

`Config.RuleApplies()` in `config.go` duplicates the logic in `Evaluator.EvaluateForStudent()` in `scorerule.go`. The config method is only called via `CalculateTotalEffect` (used in `users.go` for the user list/CSV). This means the same rule can produce different results from two different code paths — a divergence bug waiting to happen.

**Recommendation:** Remove `RuleApplies` and `CalculateScoreEffects`/`CalculateTotalEffect` from `config.go`. Use the `Evaluator` everywhere.

**3. N+1 query problem in user list**

In `UserListHandler`, for each student, `getCheckedTime` is called per rule per task, and each call does `DB.ListTaskRecords`. For a class of 30 students with 6 rules × 4 tasks, that's potentially 720 DB reads on every page load. Consider batching — load all records for a student once, then filter in memory.

Same pattern repeats in `UserInfoHandler` and `UserListCSVHandler`.

**4. `time.Now()` called in business logic makes testing impossible**

`RuleApplies` and `EvaluateForStudent` both call `time.Now()` internally. This makes the rule evaluation non-deterministic and untestable. Pass `now` as a parameter instead.

---

### Medium Issues

**5. Error silently swallowed in evaluator**

In `scorerule.go:39-41`, when `getCheckedTime` returns an error, the task is silently skipped. In `config.go`, same pattern but with a comment "Error equals task was not checked" — which is misleading since actual DB errors are also swallowed.

**6. Unused fields added in the diff**

- `TaskViewModel.CompletedCount` is added to `task.go` but `completedCount` is always 0 — never assigned.
- `TaskViewModel.CurrentTime` is added but never used in `task.html`.
- `User.RuleApplies` is set but only accessed in the template via `$.RuleApplies` — yet the template already has `.Status` on `ScoreRuleWithStatus`, making `RuleApplies` redundant.

**7. `hasTask` check is redundant**

In `user.go`, the loop checks if a score rule's task IDs exist in `AppConfig.Tasks`. But `LoadConfig` already validates this — it returns an error if a rule references a non-existent task. This check is dead code.

**8. `GetCompletedTasksCountInGroup` in `db.go` is unused**

The new method added to `storage/db.go` is never called anywhere.

---

### Minor / Style

**9. Template formatting changes are noisy** — The PR mixes feature work with whitespace reformatting (`{{ define` vs `{{define`, etc.) across multiple templates. These should be in a separate commit.

**10. Template indentation is broken in `user.html`** — The score rules table progressively loses indentation towards the bottom. Closing `</table>`, `</div>`, `</section>` are not at the correct indent level.

**11. `docker-compose.yaml` hardcodes JWT secret** — `SLAP_JWT_SECRET=test` in the main compose file; consider `docker-compose.override.yaml` or `.env` instead.

**12. Double `---` separator in GUIDE.md** — produces a redundant horizontal rule.

---

### No Tests

The PR adds significant business logic (rule evaluation with multiple condition branches) but includes zero tests — no unit tests for `RuleApplies`/`EvaluateForStudent`, no hurl integration tests.

---

### Summary

| Category | Count |
|----------|-------|
| Critical | 4 |
| Medium | 4 |
| Minor | 4 |

**Key asks before merge:**
1. Remove duplicate logic — single code path for rule evaluation
2. Add unit tests for rule evaluation (especially edge cases around `time.Now()`)
3. Add hurl tests for the score rules UI
4. Fix the N+1 query pattern or add a TODO acknowledging it
5. Remove dead code (`CalculateScoreEffects`, `CompletedCount`, `GetCompletedTasksCountInGroup`)

**Comment by ryukzak** [comment]
## Follow-up review notes

A few items from the latest pass.

### 1. Nil dereference in `min_checked_before` branch

`src/handlers/scorerule.go`:

```go
if rule.Condition.MinCheckedBefore > 0 {
    for _, t := range checkedTimes {
        if t != nil && t.Before(*rule.Condition.CheckedBefore) { // deref
            ...
        }
    }
    ...
    eval.IsActive = !now.After(*rule.Condition.CheckedBefore) // deref
```

`Condition.CheckedBefore` is `*time.Time`. `LoadConfig` doesn't reject `min_checked_before > 0` without `checked_before`, so a misconfigured rule panics on first evaluation. Either validate at load time or guard at eval time.

### 2. Errors silently swallowed

Three spots discard real DB errors:

- `src/handlers/user.go` — `eval, _ := evaluator.EvaluateForStudent(...)` in `UserInfoHandler`
- `src/handlers/user.go` — same pattern in `calculateStudentTotalEffectWithRecords`
- `src/storage/db.go` — `GetAllTaskRecordsForUser` does `continue` on `json.Unmarshal` failure with no log

A DB error quietly produces a zero `Evaluation{}`; the student sees a wrong total with no trace. At minimum log; ideally propagate.

### 3. `LoadConfig` validation is incomplete

`src/config/config.go`:

- Doesn't reject `min_checked_before > 0` without `checked_before` (root cause of #1).
- Doesn't reject `effect == 0` (silent no-op rule).
- Doesn't reject `checked_after >= checked_before` for interval rules.
- `taskExists` map is rebuilt inside the per-rule loop — hoist it.

### 4. Branch order in `EvaluateForStudent` is fragile

The four branches in `scorerule.go` are mutually exclusive only by convention:

1. `MinCheckedBefore > 0`
2. `CheckedAfter != nil && CheckedBefore == nil`
3. `CheckedBefore != nil && MinCheckedBefore == 0`
4. `CheckedAfter != nil && CheckedBefore != nil`

If a config sets both `min_checked_before` and `checked_after`, the min branch wins and `checked_after` is silently ignored. Either validate "exactly one of these shapes" in `LoadConfig`, or dispatch on an explicit `condition_type` field.

**Review by ryukzak — APPROVED** [review]
